### PR TITLE
withSlots: refactor out redundant React.Children traversal

### PR DESCRIPTION
**Pull request by KevinTCoughlin**
Visible while profiling `<Stack />`. Avoid traversing React.Children twice for children count, re-use `toArray`'s traversal.

image

Ideally this wrapper could re-use children already being traversed within `<Stack />` itself.

image

### CODE REVIEW

File: packages/foundation/src/slots.tsx
Comment (Ridvanovskyy) [inline packages/foundation/src/slots.tsx:42]: Hello, @KevinTCoughlin. I have small questions:
1. Can we take out `children = React.Children.toArray (children);` outside of check `if (slotType.isSlot) {}`? As I see this array is needed both `if` and for `else`.
2.  After we can add - **const** children = React.Children.toArray(children);

**Review by Ridvanovskyy — APPROVED** [review]

**Review by khmakoto — APPROVED** [review]

**Comment by khmakoto** [comment]
Let's wait until version 8 beta release to merge this. Alternatively, we can merge into the 7.0 branch and port into 8.0 later. See #15222 for more details.

**Review by khmakoto — CHANGES_REQUESTED** [review]
Let's wait until version 8 beta release to merge this. Alternatively, we can merge into the 7.0 branch and port into 8.0 later. See #15222 for more details. Otherwise this LGTM.

**Comment by KevinTCoughlin** [comment]
> Let's wait until version 8 beta release to merge this. Alternatively, we can merge into the 7.0 branch and port into 8.0 later. See #15222 for more details. Otherwise this LGTM.

@khmakoto here is the `7.0` cherry-pick, https://github.com/microsoft/fluentui/pull/15471. Can you ensure that is merged so OneDrive & SharePoint benefit soonest. Also please merge this change into `master` once the branch is stable so the perf fix is not lost. Thanks 🙇‍♂️.

**Review by cliffkoh — APPROVED** [review]

**Review by xugao — APPROVED** [review]
This is very interesting/good finding, thank you Kevin!

I see the gain in flamgrill tests as well:
image

As you pointed out, the `React.children.toArray` call has significant cost as well

**Review by khmakoto — APPROVED** [review]

**Pull request merged**
